passwd.1
.\" Copyright Red Hat, Inc., 1998, 1999, 2002, 2009.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, and the entire permission notice in its entirety,
.\"    including the disclaimer of warranties.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior
.\"    written permission.
.\"
.\" ALTERNATIVELY, this product may be distributed under the terms of
.\" the GNU Public License, in which case the provisions of the GPL are
.\" required INSTEAD OF the above restrictions.  (This clause is
.\" necessary due to a potential bad interaction between the GPL and
.\" the restrictions contained in a BSD-style copyright.)
.\"
.\"  THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
.\" WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
.\" OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" Copyright (c) Cristian Gafton, 1998, <gafton@redhat.com>
.\" Copyright (c) Tomas Mraz, 2009, <tmraz@redhat.com>
.\"
.TH PASSWD 1 "Sep 24 2009" "GNU/Linux" "User utilities"
.SH NAME

passwd \- update user's authentication tokens

.SH SYNOPSIS
.B passwd [-k] [-l] [-u [-f]] [-d] [-n mindays] [-x maxdays] [-w warndays] [-i inactivedays] [-S] [--stdin] [username]
.sp 2
.SH DESCRIPTION
The passwd utility is used to update user's authentication token(s).

This task is achieved through calls to the
.BR "Linux-PAM" " and "
.BR "Libuser API" ". "
Essentially, it initializes itself as a "passwd" service with
.I Linux-PAM
and utilizes configured
.I "password"
modules to authenticate and then update a user's password.

.sp
A simple entry in the global
.I Linux-PAM
configuration file for this service would be:
.br
  
.br
 #
.br
 # passwd service entry that does strength checking of
.br
 # a proposed password before updating it.
.br
 #
.br
 passwd password requisite pam_cracklib.so retry=3
.br
 passwd password required pam_unix.so use_authtok
.br
 #

.sp
Note, other module types are not required for this application to
function correctly.

.SH OPTIONS

.IP \fB-k\fR
The option
.BR -k ", "
is used to indicate that the update should only be for expired
authentication tokens (passwords); the user wishes to keep their
non-expired tokens as before.

.IP \fB-l\fR
This option is used to lock the specified account and it is available
to root only. The locking is performed by rendering the encrypted
password into an invalid string (by prefixing the encrypted string
with an !).

.IP \fB--stdin\fR
This option is used to indicate that \fBpasswd\fR should read the new
password from standard input, which can be a pipe.

.IP \fB-u\fR
This is the reverse of the
.BR -l " option - it will unlock the account"
password by removing the ! prefix. This option is available to root
only. By default passwd will refuse to create a passwordless account 
(it will not unlock an account that has only "!" as a password). The 
force option \fB-f\fR will override this protection.

.IP \fB-d\fR
This is a quick way to delete a password for an account. It will set
the named account passwordless. Available to root only.

.IP \fB-n\fR
This will set the minimum password lifetime, in days, if the user's
account supports password lifetimes.  Available to root only.

.IP \fB-x\fR
This will set the maximum password lifetime, in days, if the user's
account supports password lifetimes.  Available to root only.

.IP \fB-w\fR
This will set the number of days in advance the user will begin receiving
warnings that her password will expire, if the user's account supports
password lifetimes.  Available to root only.

.IP \fB-i\fR
This will set the number of days which will pass before an expired password
for this account will be taken to mean that the account is inactive and should
be disabled, if the user's account supports password lifetimes.  Available to
root only.

.IP \fB-S\fR
This will output a short information about the status of the password
for a given account. Available to root user only.

.SH "Remember the following two principles"

.IP \fBProtect\ your\ password.\fR
Don't write down your password - memorize it.
In particular, don't write it down and leave it anywhere, and don't
place it in an unencrypted file!  Use unrelated passwords for
systems controlled by different organizations.  Don't give or share your
password, in particular to someone claiming to be from
computer support or a vendor.  Don't let anyone watch you enter your
password.  Don't enter your password to a computer you don't trust or
if things "look funny"; someone may be trying to hijack your password.
Use the password for a limited time and change it periodically.

.IP \fBChoose\ a\ hard-to-guess\ password.\fR
.I passwd
through the calls to the
.BR pam_cracklib " PAM module"
will try to prevent you from choosing a really bad password,
but it isn't foolproof; create your password wisely.
Don't use something you'd find in a dictionary (in any language or jargon).
Don't use a name (including that of a spouse, parent, child, pet,
fantasy character, famous person, and location) or any
variation of your personal or account name.  Don't use accessible
information about you (such as your phone number, license plate, or
social security number) or your environment.  Don't use a birthday or a
simple pattern (such as "qwerty", "abc", or "aaa").  Don't use any of those
backwards, followed by a digit, or preceded by a digit. Instead, use
a mixture of upper and lower case letters, as well as digits or
punctuation.  When choosing a new password, make sure it's unrelated
to any previous password. Use long passwords (say at least 8 characters
long).  You might use a word pair with punctuation inserted, a
passphrase (an understandable sequence of words), or the first
letter of each word in a passphrase.

.SH ""
These principles are partially enforced by the system, but only partly so.
Vigilence on your part will make the system much more secure.

.SH "EXIT CODE"

On successful completion of its task,
.B passwd
will complete with exit code 0.  An exit code of 1 indicates an error
occurred.  Textual errors are written to the standard error stream.

.SH "CONFORMING TO"
.br
.BR Linux-PAM
(Pluggable Authentication modules for Linux).

.SH "FILES"
.br
.B /etc/pam.d/passwd
- the
.BR Linux-PAM
configuration file

.SH BUGS
.sp 2
None known.

.SH "SEE ALSO"

.BR pam "(8), "
.BR pam.d "(5), "
.BR libuser.conf "(5), "
and
.BR pam_chauthtok "(3). "

.sp
For more complete information on how to configure this application
with
.BR Linux-PAM ", "
see the
.BR "Linux-PAM System Administrators' Guide" "."

.SH AUTHOR
Cristian Gafton <gafton@redhat.com>


passwd.c
/*
 * Copyright Red Hat, Inc., 1998, 1999, 2001, 2002.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, and the entire permission notice in its entirety,
 *    including the disclaimer of warranties.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 *
 * ALTERNATIVELY, this product may be distributed under the terms of
 * the GNU Public License, in which case the provisions of the GPL are
 * required INSTEAD OF the above restrictions.  (This clause is
 * necessary due to a potential bad interaction between the GPL and
 * the restrictions contained in a BSD-style copyright.)
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* Written by Cristian Gafton <gafton@redhat.com> */

#include "config.h"

#include <sys/types.h>

#include <ctype.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <popt.h>
#include <errno.h>
#include <limits.h>
#include <syslog.h>
#include <locale.h>
#include <libintl.h>
#include "pwdb.h"

#ifdef WITH_SELINUX
#include <selinux/selinux.h>
#include <selinux/context.h>
#include <selinux/av_permissions.h>
#include "selinux_utils.h"
#endif

#ifdef WITH_AUDIT
#include <libaudit.h>
#else
#define audit_log_user_message(d,ty,m,h,a,t,r) do { ; } while(0) 
#define audit_log_acct_message(d,ty,p,o,n,i,h,a,t,r) do { ; } while(0) 
static int audit_open(void) { errno = EPROTONOSUPPORT; return -1; } 
#endif

#include <security/pam_appl.h>
#include <security/pam_misc.h>

static int audit_fd = -1;

/* conversation function & corresponding structure */
static struct pam_conv conv = {
	misc_conv,
	NULL
};

const char *username = NULL;	/* username specified on the command line */
const char *progname = NULL;	/* the name of the program */
int passwd_flags = 0;		/* flags specified by root */

#define PASSWD_KEEP	0x0001	/* keep un-expired tokens */

#define PASSWD_LOCK	0x0002	/* lock the password */
#define PASSWD_UNLOCK	0x0004	/* unlock the password, if locked */
#define PASSWD_DELETE	0x0008	/* delete the user's password */
#define PASSWD_STATUS	0x0010	/* report the password status */
#define PASSWD_FORCE	0x0020	/* force change of expired token */
#define PASSWD_STDIN	0x0040	/* read the password from stdin (root only) */
#define PASSWD_ROOT	0x001E	/* options which are mutually exclusive */

#define PASSWD_MIN	0x0100	/* set the minimum password lifetime */
#define PASSWD_MAX	0x0200	/* set the maximum password lifetime */
#define PASSWD_WARN	0x0400	/* set the password warning */
#define PASSWD_INACT	0x0800	/* set the inactive time */
#define PASSWD_AGING	0x0F00	/* aging options */

#ifdef HAVE_PAM_FAIL_DELAY
#define PASSWD_FAIL_DELAY	2000000	/* usec delay on failure */
#endif


/* A conversation function which uses an internally-stored value for
 * the responses. */
static int
stdin_conv(int num_msg, const struct pam_message **msgm,
	   struct pam_response **response, void *appdata_ptr)
{
	struct pam_response *reply;
	int count;

	/* Sanity test. */
	if (num_msg <= 0) {
		return PAM_CONV_ERR;
	}

	/* Allocate memory for the responses. */
	reply = calloc(num_msg, sizeof(struct pam_response));
	if (reply == NULL) {
		return PAM_CONV_ERR;
	}

	/* Each prompt elicits the same response. */
	for (count = 0; count < num_msg; ++count) {
		if (msgm[count]->msg_style == PAM_PROMPT_ECHO_OFF) {
			reply[count].resp_retcode = 0;
			reply[count].resp = strdup(appdata_ptr);
		} else {
			reply[count].resp_retcode = 0;
			reply[count].resp = strdup("");
		}
	}

	/* Set the pointers in the response structure and return. */
	*response = reply;
	return PAM_SUCCESS;
}

/* Parse command-line arguments, rejecting conflicting flags and performing
 * various other initialization tasks. */
static void
parse_args(int argc, const char **argv,
	   long *min, long *max, long *warn, long *inact)
{
	poptContext optCon;
	int delete = 0, force = 0, keep = 0, lock = 0, status = 0, unlock = 0;
	int use_stdin = 0;
	int rc;
	const char **extraArgs;
	struct poptOption options[] = {
		{"keep-tokens", 'k', POPT_ARG_NONE, &keep, 0,
		 _("keep non-expired authentication tokens")},
		{"delete", 'd', POPT_ARG_NONE, &delete, 0,
		 _("delete the password for the named account (root only)")},
		{"lock", 'l', POPT_ARG_NONE, &lock, 0,
		 _("lock the named account (root only)")},
		{"unlock", 'u', POPT_ARG_NONE, &unlock, 0,
		 _("unlock the named account (root only)")},
		{"force", 'f', POPT_ARG_NONE, &force, 0,
		 _("force operation")},
		{"maximum", 'x', POPT_ARG_LONG, max, 0,
		 _("maximum password lifetime (root only)"), "DAYS"},
		{"minimum", 'n', POPT_ARG_LONG, min, 0,
		 _("minimum password lifetime (root only)"), "DAYS"},
		{"warning", 'w', POPT_ARG_LONG, warn, 0,
		 _("number of days warning users receives before password "
		 "expiration (root only)"), "DAYS"},
		{"inactive", 'i', POPT_ARG_LONG, inact, 0,
		 _("number of days after password expiration when an account "
		 "becomes disabled (root only)"), "DAYS"},
		{"status", 'S', POPT_ARG_NONE, &status, 0,
		 _("report password status on the named account (root only)")},
		{"stdin", '\0', POPT_ARG_NONE, &use_stdin, 0,
		 _("read new tokens from stdin (root only)")},
		POPT_AUTOHELP {NULL, '\0', 0, NULL},
	};
	struct passwd *pw;

	*min = *max = *warn = *inact = -2;
	optCon = poptGetContext("passwd", argc, argv, options, 0);
	poptSetOtherOptionHelp(optCon, _("[OPTION...] <accountName>"));

	if ((rc = poptGetNextOpt(optCon)) < -1) {
		fprintf(stderr, _("%s: bad argument %s: %s\n"), progname,
			poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
			poptStrerror(rc));
		exit(-3);
	}

	extraArgs = poptGetArgs(optCon);

	if (keep) {
		passwd_flags |= PASSWD_KEEP;
	}
	if (lock) {
		passwd_flags |= PASSWD_LOCK;
	}
	if (unlock) {
		passwd_flags |= PASSWD_UNLOCK;
	}
	if (status) {
		passwd_flags |= PASSWD_STATUS;
	}
	if (delete) {
		passwd_flags |= PASSWD_DELETE;
	}
	if (force) {
		passwd_flags |= PASSWD_FORCE;
	}
	if (use_stdin) {
		passwd_flags |= PASSWD_STDIN;
	}
	if (*min != -2) {
		passwd_flags |= PASSWD_MIN;
	}
	if (*max != -2) {
		passwd_flags |= PASSWD_MAX;
	}
	if (*warn != -2) {
		passwd_flags |= PASSWD_WARN;
	}
	if (*inact != -2) {
		passwd_flags |= PASSWD_INACT;
	}

	/* The rest of the flags are mutually-exclusive, except for --force. */
	if (passwd_flags) {
		int tmp;
		int count;
		tmp = passwd_flags & PASSWD_ROOT;
		count = 0;
		/* Check the rightmost bit and shift right. */
		while (tmp != 0) {
			if (tmp & 0x01) {
				count++;
			}
			tmp = tmp >> 1;
		}
		/* Error if other bits are set. */
		if (count > 1) {
			fprintf(stderr,
				_("%s: Only one of -l, -u, -d, -S may be specified.\n"),
				progname);
			exit(-2);
		}
		/* Error out if we had -l/-u/-d/-S and an aging option. */
		if (count > 0) {
			tmp = passwd_flags & PASSWD_AGING;
			if (tmp != 0) {
				fprintf(stderr,
					_("%s: Cannot mix one of -l, -u, -d, -S and one of -i, -n, -w, -x.\n"),
					progname);
				exit(-2);
			}
		}
	}

	/* The only flag which unprivileged users get to use is -k. */
	if ((passwd_flags & ~PASSWD_KEEP) && 
#ifdef WITH_SELINUX
	    ((getuid() != 0) || (is_selinux_enabled() > 0 &&
		    checkPasswdAccess(PASSWD__PASSWD)))) {
#else
	    (getuid() != 0)) {
#endif
		if (passwd_flags & PASSWD_STATUS) {
			audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
				NULL, "password status display",
				NULL, getuid(), NULL, NULL, NULL, 0);
		} else {
			audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
				NULL, "password attribute change",
				NULL, getuid(), NULL, NULL, NULL, 0);
		}
		fprintf(stderr, _("Only root can do that.\n"));
		exit(-2);
	}

	/* Only root gets to specify a user name. */
	username = NULL;
	if ((extraArgs != NULL) && (extraArgs[0] != NULL)) {
		if (getuid() != 0) {
			/* The invoking user was not root. */
			audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
				NULL, "password change", extraArgs[0],
				getuid(), NULL, NULL, NULL, 0);
			fprintf(stderr,
				_("%s: Only root can specify a user name.\n"),
				progname);
			exit(-3);
		} else {
			/* The invoking user was root. */
			username = extraArgs[0];
			/* Sanity-check the user name */
			if (strlen(username) > MAX_USERNAMESIZE) {
				fprintf(stderr,
					_("%s: The user name supplied is too long.\n"),
					progname);
				exit(-3);
			}
		}

		/* If there is more than one unrecognized argument, we suddenly
		 * get confused. */
		if (extraArgs[1] != NULL) {
			fprintf(stderr,
				_("%s: Only one user name may be specified.\n"),
				progname);
			exit(-3);
		}
	}

	/* Now if any of the -l, -u, -d, -S, -i, -n, -w, or -x options were
	 * given, and a username was not specified, bail out. */
	if ((passwd_flags & ~PASSWD_KEEP) && (username == NULL)) {
		fprintf(stderr,
			_("%s: This option requires a user name.\n"),
			progname);
		exit(-2);
	}

	/* Determine the name of the user whose account we're operating on,
	 * and make sure the account exists. */
	if (username == NULL) {
		/* The invoking user. */
		pw = getpwuid(getuid());
		if (pw == NULL) {
			fprintf(stderr, _("%s: Can not identify you!\n"),
				progname);
			exit(-3);
		}
		username = strdup(pw->pw_name);
	} else {
		/* The name specified on the command-line. */
		pw = getpwnam(username);
		if (pw == NULL) {
			fprintf(stderr, _("%s: Unknown user name '%s'.\n"),
				progname, username);
			exit(-4);
		}
	}
}

int
main(int argc, const char **argv)
{
	int retval;
	long min, max, warn, inact;
	pam_handle_t *pamh = NULL;
	struct passwd *pwd;
	char *tty_name, *ttyn;

	setlocale(LC_ALL, "");
	bindtextdomain("passwd", "/usr/share/locale");
	textdomain("passwd");
	
	audit_fd = audit_open();
	if (audit_fd < 0 && !(errno == EINVAL || errno == EPROTONOSUPPORT ||
				errno == EAFNOSUPPORT)) {
		/* The above error codes are only given when the kernel doesn't
		 * have audit compiled in. */
		fprintf(stderr, "Error - unable to connect to audit system\n");
		exit(1);
	}

	/* Parse command-line arguments. */
	progname = basename(argv[0]);
	parse_args(argc, argv, &min, &max, &warn, &inact);

	pwd = getpwnam(username);
	if (pwd == NULL) {
		fprintf(stderr, _("%s: Unknown user name '%s'.\n"),
			progname, username);
		exit(-4);
	}

#ifdef WITH_SELINUX
	if ((is_selinux_enabled() > 0) &&
	    (getuid() == 0) &&
	    (check_selinux_access(username, pwd->pw_uid, PASSWD__PASSWD) != 0)) {
		security_context_t user_context;
		if (getprevcon(&user_context) < 0) {
			user_context = strdup("Unknown user context");
		}
		syslog(LOG_ALERT,
		       "%s is not authorized to change the password of %s",
		       user_context, username);
		fprintf(stderr,
			_("%s: %s is not authorized to change the "
			  "password of %s\n"),
			progname, user_context, username);
		freecon(user_context);
		audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
			NULL, "change password", NULL, pwd->pw_uid,
			NULL, NULL, NULL, 0);
		exit(1);
	}
#endif

	/* Handle account locking request. */
	if (passwd_flags & PASSWD_LOCK) {
		printf(_("Locking password for user %s.\n"), username);
		retval = pwdb_lock_password(username);
		printf("%s: %s\n", progname,
		       retval ==
		       0 ? _("Success") : _("Error (password not set?)"));
		audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
			NULL, "lock password", NULL, pwd->pw_uid,
			NULL, NULL, NULL, retval == 0);
		return retval;
	}
	/* Handle account unlocking request. */
	if (passwd_flags & PASSWD_UNLOCK) {
		printf(_("Unlocking password for user %s.\n"), username);
		retval = pwdb_unlock_password(username,
					      passwd_flags & PASSWD_FORCE);
		printf("%s: %s\n", progname,
		       retval == 0 ? _("Success") :
		       retval ==
		       -2 ? _("Unsafe operation (use -f to force)") :
		       _("Error (password not set?)"));
		audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
			NULL, "unlock password", NULL, pwd->pw_uid,
			NULL, NULL, NULL, retval == 0);
		return retval;
	}
	/* Handle password clearing request. */
	if (passwd_flags & PASSWD_DELETE) {
		printf(_("Removing password for user %s.\n"), username);
		retval = pwdb_clear_password(username);
		printf("%s: %s\n", progname,
		       (retval == 0) ? _("Success") : _("Error"));
		audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
			NULL, "delete password", NULL, pwd->pw_uid,
			NULL, NULL, NULL, retval == 0);
		return retval;
	}
	/* Display account status. */
	if (passwd_flags & PASSWD_STATUS) {
		retval = pwdb_display_status(username);
		audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
			NULL, "password status displayed for user",
			NULL, pwd->pw_uid, NULL, NULL, NULL, retval == 0);
		return retval;
	}
	/* Adjust aging parameters. */
	if (passwd_flags & PASSWD_AGING) {
		char aubuf[PATH_MAX];
		printf(_("Adjusting aging data for user %s.\n"), username);
		retval = pwdb_update_aging(username, min, max, warn, inact);
		printf("%s: %s\n", progname,
		       (retval == 0) ? _("Success") : _("Error"));
		snprintf(aubuf, sizeof(aubuf), "password aging data updated "
				"- acct=%s, uid=%u, min=%li, max=%li,"
				" warn=%li, inact=%li", username, 
				pwd->pw_uid, min, max, warn, inact);
		audit_log_user_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
			aubuf, NULL, NULL, NULL, retval == 0);
		return retval;
	}

	/* The standard behavior follows.  At this point we know for whom
	 * we are going to change a password, so let the invoking user
	 * know what's going on. */
	printf(_("Changing password for user %s.\n"), username);

	/* If we need to read the new password from stdin, read it and switch
	 * to the really-quiet stdin conversation function. */
	if (passwd_flags & PASSWD_STDIN) {
		char *ptr, newPassword[80];
		int i;

		i = read(STDIN_FILENO, newPassword,
			 sizeof(newPassword) - 1);
		if (i < 0) {
			fprintf(stderr,
				_("%s: error reading from stdin: %s\n"), progname,
				strerror(errno));
			exit(1);
		}

		newPassword[i] = '\0';
		ptr = strchr(newPassword, '\n');
		if (ptr)
			*ptr = 0;
		conv.conv = stdin_conv;
		conv.appdata_ptr = strdup(newPassword);
	}

	/* Start up PAM. */
	retval = pam_start("passwd", username, &conv, &pamh);
	if (retval != PAM_SUCCESS) {
		fprintf(stderr,
			_("%s: unable to start pam: %s\n"), progname,
			pam_strerror(pamh, retval));
		exit(1);
	}

	if ((ttyn = ttyname(0)) != NULL) {	
		if (strncmp(ttyn, "/dev/", 5) == 0)
			tty_name = ttyn+5;
		else
			tty_name = ttyn;
		retval = pam_set_item(pamh, PAM_TTY, tty_name);
		if (retval != PAM_SUCCESS) {
			fprintf(stderr,
				_("%s: unable to set tty for pam: %s\n"), progname,
				pam_strerror(pamh, retval));
			pam_end(pamh, retval);
			exit(1);
		}
	}
#ifdef HAVE_PAM_FAIL_DELAY
	/* We have to pause on failure, so tell libpam the minimum amount
	 * of time it should wait after a failure. */
	retval = pam_fail_delay(pamh, PASSWD_FAIL_DELAY);
	if (retval != PAM_SUCCESS) {
		fprintf(stderr, _("%s: unable to set failure delay: %s\n"),
			progname, pam_strerror(pamh, retval));
		exit(1);
	}
#endif

	/* Go for it. */
	retval = pam_chauthtok(pamh,
			       (passwd_flags & PASSWD_KEEP) ?
			       PAM_CHANGE_EXPIRED_AUTHTOK : 0);
	if (retval == PAM_SUCCESS) {
		/* We're done.  Tell the invoking user that it worked. */
		retval = pam_end(pamh, PAM_SUCCESS);
		if (passwd_flags & PASSWD_KEEP) {
			audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
				NULL, "change expired password", NULL,
				pwd->pw_uid, NULL, NULL, NULL,
				retval == PAM_SUCCESS);
			printf(_("%s: expired authentication tokens updated successfully.\n"),
				progname);
		} else {
			audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
				NULL, "change password", NULL, pwd->pw_uid,
				NULL, NULL, NULL, retval == PAM_SUCCESS);
			printf(_("%s: all authentication tokens updated successfully.\n"),
				progname);
		}
		retval = 0;
	} else {
		/* Horrors!  It failed.  Relay the bad news. */
		audit_log_acct_message(audit_fd,  AUDIT_USER_CHAUTHTOK,
				NULL, "change password", NULL, pwd->pw_uid,
				NULL, NULL, NULL, retval == PAM_SUCCESS);
		fprintf(stderr, "%s: %s\n", progname,
			pam_strerror(pamh, retval));
		pam_end(pamh, retval);
		retval = 1;
	}
	return retval;
}
